# Shopify, Fall 2022 Data Science Intern Challenge 

## Data load and exploration
### Observations:
- No missing values
- 3 possible payment method type: credit_card, debit, cash
- Maximum order amount: 704000

In [3]:
import pandas as pd
import numpy as np
import seaborn as sns
sns.set()

In [15]:
# Load dataset
df = pd.read_csv('2019 Winter Data Science Intern Challenge Data Set - Sheet1.csv', parse_dates=["created_at"])
df.head()

order_id  shop_id  user_id  order_amount  total_items payment_method  \
0         1       53      746           224            2           cash   
1         2       92      925            90            1           cash   
2         3       44      861           144            1           cash   
3         4       18      935           156            1    credit_card   
4         5       18      883           156            1    credit_card   

           created_at  
0 2017-03-13 12:36:56  
1 2017-03-03 17:38:52  
2 2017-03-14 04:23:56  
3 2017-03-26 12:43:37  
4 2017-03-01 04:35:11

In [18]:
# brief data exploration
df.info()
df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   order_id        5000 non-null   int64         
 1   shop_id         5000 non-null   int64         
 2   user_id         5000 non-null   int64         
 3   order_amount    5000 non-null   int64         
 4   total_items     5000 non-null   int64         
 5   payment_method  5000 non-null   object        
 6   created_at      5000 non-null   datetime64[ns]
dtypes: datetime64[ns](1), int64(5), object(1)
memory usage: 254.0+ KB


order_id      shop_id      user_id   order_amount  total_items
count  5000.000000  5000.000000  5000.000000    5000.000000   5000.00000
mean   2500.500000    50.078800   849.092400    3145.128000      8.78720
std    1443.520003    29.006118    87.798982   41282.539349    116.32032
min       1.000000     1.000000   607.000000      90.000000      1.00000
25%    1250.750000    24.000000   775.000000     163.000000      1.00000
50%    2500.500000    50.000000   849.000000     284.000000      2.00000
75%    3750.250000    75.000000   925.000000     390.000000      3.00000
max    5000.000000   100.000000   999.000000  704000.000000   2000.00000

In [19]:
pd.value_counts(df.payment_method)

credit_card    1735
debit          1671
cash           1594
Name: payment_method, dtype: int64

## Question 1:

### a. Since the value of the AOV is highly unusual I start by looking for outliers, perhaps caused by entry errors since the max value of an order is 704000.
(Let's define outliers, in this case, as the values above the 99 percentile)

In [28]:
order_amount_3rd_percentile = np.percentile(df["order_amount"], 99)
order_amount_3rd_percentile

25725.0

In [29]:
df[df["order_amount"]>order_amount_3rd_percentile]

order_id  shop_id  user_id  order_amount  total_items payment_method  \
15          16       42      607        704000         2000    credit_card   
60          61       42      607        704000         2000    credit_card   
490        491       78      936         51450            2          debit   
493        494       78      983         51450            2           cash   
511        512       78      967         51450            2           cash   
520        521       42      607        704000         2000    credit_card   
617        618       78      760         51450            2           cash   
691        692       78      878        154350            6          debit   
1104      1105       42      607        704000         2000    credit_card   
1259      1260       78      775         77175            3    credit_card   
1362      1363       42      607        704000         2000    credit_card   
1436      1437       42      607        704000         2000    credit_card   
1529      1530       78      810         51450            2           cash   
1562      1563       42      607        704000         2000    credit_card   
1602      1603       42      607        704000         2000    credit_card   
2153      2154       42      607        704000         2000    credit_card   
2297      2298       42      607        704000         2000    credit_card   
2452      2453       78      709         51450            2           cash   
2492      2493       78      834        102900            4          debit   
2495      2496       78      707         51450            2           cash   
2512      2513       78      935         51450            2          debit   
2564      2565       78      915         77175            3          debit   
2690      2691       78      962         77175            3          debit   
2818      2819       78      869         51450            2          debit   
2821      2822       78      814         51450            2           cash   
2835      2836       42      607        704000         2000    credit_card   
2906      2907       78      817         77175            3          debit   
2969      2970       42      607        704000         2000    credit_card   
3101      3102       78      855         51450            2    credit_card   
3167      3168       78      927         51450            2           cash   
3332      3333       42      607        704000         2000    credit_card   
3403      3404       78      928         77175            3          debit   
3705      3706       78      828         51450            2    credit_card   
3724      3725       78      766         77175            3    credit_card   
4056      4057       42      607        704000         2000    credit_card   
4079      4080       78      946         51450            2           cash   
4192      4193       78      787         77175            3    credit_card   
4311      4312       78      960         51450            2          debit   
4412      4413       78      756         51450            2          debit   
4420      4421       78      969         77175            3          debit   
4646      4647       42      607        704000         2000    credit_card   
4715      4716       78      818         77175            3          debit   
4868      4869       42      607        704000         2000    credit_card   
4882      4883       42      607        704000         2000    credit_card   

              created_at  
15   2017-03-07 04:00:00  
60   2017-03-04 04:00:00  
490  2017-03-26 17:08:19  
493  2017-03-16 21:39:35  
511  2017-03-09 07:23:14  
520  2017-03-02 04:00:00  
617  2017-03-18 11:18:42  
691  2017-03-27 22:51:43  
1104 2017-03-24 04:00:00  
1259 2017-03-27 09:27:20  
1362 2017-03-15 04:00:00  
1436 2017-03-11 04:00:00  
1529 2017-03-29 07:12:01  
1562 2017-03-19 04:00:00  
1602 2017-03-17 04:00:00  
2153 2017-03-12 04:00:00  
2297 2017-03-07 04:00:00  
2452 2017

##### Given this table, it comes out that there are several rows with anomalous values. In particular the shop with the id:42, inserted many equal values with order_amount of 704000, total_items 2000, at 4 am. This leads me to think that there is a problem in the computer system of this shop

### b. In these cases it is preferable to use as metric the median instead of the mean, that is useful in case of dataset with evident outilers
### c. The value of the median of this dataset for the column "order_amount" is 284. Previously calculate in the dataset report